// kittik-shape-image/lib/Image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _kittikShapeBasic = require('kittik-shape-basic');

var _kittikShapeBasic2 = _interopRequireDefault(_kittikShapeBasic);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Implements support for Image drawing in terminal.
 *
 * @extends {Shape}
 * @since 1.0.0
 */
class Image extends _kittikShapeBasic2.default {
  /**
   * Creates new Image instance.
   * Worth noting, that Image can be rendered only in iTerm 3.
   * Applies immediately, without caching in virtual terminal in {@link Cursor}.
   *
   * @constructor
   * @param {Cursor} cursor Cursor instance
   * @param {Object} [options] Options object
   * @param {String} [options.image] Base64 encoded file or path to image
   * @param {Boolean} [options.preserveAspectRatio] If true, preserve aspect ratio
   * @example
   * Image.create(cursor, {
   *   x: 'center',
   *   y: 'middle',
   *   image: 'my-image.png',
   *   preserveAspectRatio: true
   * });
   */
  constructor(cursor) {
    let options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    super(cursor, options);

    this.setImage(options.image);
    this.setPreserveAspectRatio(options.preserveAspectRatio);
  }

  /**
   * Get base64 encoded image.
   *
   * @returns {String}
   */
  getImage() {
    return this.get('image');
  }

  /**
   * Set image to the shape.
   *
   * @param {String} image Can be path to the image or base64 encoded image
   * @returns {Image}
   * @example
   * image.setShape('base64');
   * image.setShape('./my-image.png');
   */
  setImage(image) {
    if (!image) return this;
    if (Image.isBase64(image)) return this.set('image', image);

    return this.set('image', _fs2.default.readFileSync(_path2.default.resolve(image), 'base64'));
  }

  /**
   * Check if image is preserve aspect ratio.
   *
   * @returns {Boolean}
   */
  isPreserveAspectRatio() {
    return !!this.get('preserveAspectRatio');
  }

  /**
   * Set preserve aspect ratio.
   *
   * @param {Boolean} [isPreserveAspectRatio=true] If false, not preserves aspect ratio in the image
   * @returns {Image}
   */
  setPreserveAspectRatio() {
    let isPreserveAspectRatio = arguments.length <= 0 || arguments[0] === undefined ? true : arguments[0];

    return this.set('preserveAspectRatio', isPreserveAspectRatio);
  }

  /**
   * Renders the shape.
   *
   * @override
   * @returns {Image}
   */
  render() {
    const cursor = this.getCursor();
    const width = this.getWidth();
    const height = this.getHeight();
    const x = this.getX();
    const y = this.getY();
    const image = this.getImage();
    const preserveAspectRatio = this.isPreserveAspectRatio();
    const args = `width=${ width };height=${ height };preserveAspectRatio=${ preserveAspectRatio ? 1 : 0 };inline=1`;

    cursor._stream.write(`\u001b[${ y + 1 };${ x + 1 }H\u001b]1337;File=${ args }:${ image }^G`);

    return this;
  }

  /**
   * Serializes shape to the object representation.
   *
   * @returns {Object}
   */
  toObject() {
    const obj = super.toObject();

    Object.assign(obj.options, {
      image: this.get('image'),
      preserveAspectRatio: this.get('preserveAspectRatio')
    });

    return obj;
  }

  /**
   * Check if string is base64 encoded string.
   *
   * @static
   * @param {String} string
   * @returns {Boolean}
   */
  static isBase64(string) {
    return (/^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/.test(string)
    );
  }
}
exports.default = Image;
module.exports = exports['default'];
