// kittik-shape-code/lib/Code.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _kittikShapeBasic = require('kittik-shape-basic');

var _kittikShapeBasic2 = _interopRequireDefault(_kittikShapeBasic);

var _jsBeautify = require('js-beautify');

var _redeyed = require('redeyed');

var _redeyed2 = _interopRequireDefault(_redeyed);

var _default = require('./themes/default');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i]; return arr2; } else { return Array.from(arr); } }

/**
 * Implements Code shape which renders the highlighted code at specified point.
 *
 * @since 1.0.0
 * @extends {Shape}
 */
class Code extends _kittikShapeBasic2.default {
  /**
   * Create Code shape.
   *
   * @constructor
   * @param {Cursor} cursor Cursor instance
   * @param {Object} [options] Options object
   * @param {String} [options.code] Code fragment that you want to show
   * @example
   * Code.create(cursor, {
   *   code: 'console.log('Your code here');'
   * });
   */
  constructor(cursor) {
    let options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    super(cursor, options);

    this.setCode(options.code);
  }

  /**
   * Get current code from shape.
   *
   * @returns {String}
   */
  getCode() {
    return this.get('code');
  }

  /**
   * Set new code to this shape.
   *
   * @param {String} [code='']
   * @returns {Shape}
   */
  setCode() {
    let code = arguments.length <= 0 || arguments[0] === undefined ? '' : arguments[0];

    return this.set('code', (0, _jsBeautify.js_beautify)(code, { indent_size: 2 }));
  }

  /**
   * Returns actual width of the shape.
   * Since code shape hasn't specified width, we need to override default.
   * It looks for the longest line in your shape and returns its length.
   *
   * @override
   * @returns {Number}
   */
  getWidth() {
    const code = this.getCode().split('\n').map(item => item.length);
    return Math.max.apply(Math, _toConsumableArray(code));
  }

  /**
   * Returns actual height of the shape.
   * Since code shape hasn't specified height, we need to override default.
   * It returns count of lines in our shape.
   *
   * @override
   * @returns {Number}
   */
  getHeight() {
    return this.getCode().split('\n').length;
  }

  /**
   * Render the shape based on options.
   *
   * @override
   * @returns {Text}
   */
  render() {
    const cursor = this.getCursor();
    const codeSplits = (0, _redeyed2.default)(this.getCode(), _default.DEFAULT_THEME).splits;
    const x = this.getX();
    const y = this.getY();

    cursor.moveTo(x, y);

    codeSplits.forEach(split => {
      const code = split.replace(/__.*__/, '');
      const color = split.match(/__(.*)__/);

      if (code.match(/\n/)) return cursor.moveTo(x, cursor._y + 1).write(code.replace('\n', ''));

      cursor.foreground(color ? color[1] : 'none');
      cursor.write(code);
    });

    return this;
  }

  /**
   * Overrides default toObject() method because we have new fields here.
   *
   * @override
   * @returns {Object}
   */
  toObject() {
    const obj = super.toObject();

    Object.assign(obj.options, {
      code: this.get('code')
    });

    return obj;
  }
}
exports.default = Code;
module.exports = exports['default'];
